File: HW01/conv.c
// This is the convolutiona function in C
#include<stdio.h>

void conv (float* op, float* x, float* k, int x_row, int x_col, int k_row, int k_col);

void conv (float* op, float* x, float* k, int x_row, int x_col, int k_row, int k_col) {
        
      int conv_row = x_row-k_row+1;
      int conv_col = x_col-k_col+1;
      
      //for addressing the output image
      int row, col;
      
      //for addressing the kernel
      int i, j;
      
      //for addressing the input image
      int row_temp, col_temp;
      
      for (row = 0; row <= (conv_row-1); row++) {
         for (col = 0; col <= (conv_col-1); col++) { 
            *(op + (row * conv_col) + col) = 0;
            for (i = k_row-1; i >= 0; i--) {
               for (j = k_col-1; j >= 0; j--) {
        	       row_temp = row + k_row - (i+1);
        	       col_temp = col + k_col - (j+1);
        	       *(op + (row * conv_col + col)) += (*(x + (row_temp*x_col + col_temp))) * (*(k + (i*k_col + j)));
               }
            }
         }
      }

      return;
}

File: HW01/conv.lua
-- This is the API for 2-d convolution function written in lua

torch.setdefaulttensortype('torch.FloatTensor')
local convLib = {}

function convLib.Lua_conv(x, k)
        local x_row = x:size(1)
	local x_col = x:size(2)
	local k_row = k:size(1)
	local k_col = k:size(2)
	
	local conv_row = x_row-k_row+1
	local conv_col = x_col-k_col+1

	local op = torch.Tensor(conv_row, conv_col)
	
	for row = 1,conv_row do
	    for col = 1,conv_col do
	       op[row][col] = 0;
	       for i = k_row,1,-1 do
		  for j = k_col,1,-1 do
		     op[row][col] = op[row][col] + x[row+(k_row-i)][col+(k_col-j)] * k[i][j]
		  end
	       end
	    end
	end

	return op; 
end

-- This is the API for 2-d convolution function written in C and wrapped by lua

function convLib.C_conv(x,k)
      local x_row = x:size(1)
      local x_col = x:size(2)
      local k_row = k:size(1)
      local k_col = k:size(2)

      local conv_row = x_row-k_row+1
      local conv_col = x_col-k_col+1

      -- FFI stuff -------------------------------------------------------------------
      -- Require ffi
      ffi = require("ffi")
      -- Load myLib
      myLib = ffi.load(paths.cwd() .. '/libconv.so')
      -- Function prototypes definition
      ffi.cdef [[
      void conv(float *op, float *x, float *k, int x_row, int x_col, int k_row, int k_col)
      ]]

      -- Main program ----------------------------------------------------------------
      op = torch.Tensor(conv_row, conv_col)
      myLib.conv(torch.data(op), torch.data(x), torch.data(k), x_row, x_col, k_row, k_col)
      
      -- return the output
      return op;
end

return convLib
